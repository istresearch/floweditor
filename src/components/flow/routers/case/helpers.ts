import { isRelativeDate } from 'components/flow/routers/helpers';
import { Operator, Operators } from 'config/interfaces';
import { getOperatorConfig } from 'config/operatorConfigs';
import {
  LessThan,
  MoreThan,
  Numeric,
  NumOrExp,
  Required,
  validate,
  IsValidIntent
} from 'store/validators';
import { titleCase } from 'utils';

import { CaseElementProps, CaseElementState } from './CaseElement';
import { SelectOption } from 'components/form/select/SelectElement';
import { Asset } from 'store/flowContext';
import i18n from 'config/i18n';

export const initializeForm = (props: CaseElementProps): CaseElementState => {
  const arg1 =
    props.kase.arguments && props.kase.arguments.length >= 1 ? props.kase.arguments[0] : '';
  const arg2 =
    props.kase.arguments && props.kase.arguments.length === 2 ? props.kase.arguments[1] : '';

  return {
    errors: [],
    operatorConfig: getOperatorConfig(props.kase.type),
    argument: { value: arg1 },
    min: { value: arg1 },
    max: { value: arg2 },
    state: { value: arg1 },
    district: { value: arg2 },
    intent: { value: arg1 ? { name: arg1, value: arg1 } : null },
    confidence: { value: arg2 },
    categoryName: { value: props.categoryName || '' },
    categoryNameEdited: !!props.categoryName,
    valid: true
  };
};
/**
 * Determines prefix for case's exit name
 */
export const prefix = (operatorType: string): string => {
  let pre = '';

  if (operatorType.indexOf('_lt') > -1) {
    if (operatorType.indexOf('date') > -1) {
      pre = 'Before ';
    } else {
      if (operatorType.indexOf('lte') > -1) {
        pre = '<= ';
      } else {
        pre = '< ';
      }
    }
  } else if (operatorType.indexOf('_gt') > -1) {
    if (operatorType.indexOf('date') > -1) {
      pre = 'After ';
    } else {
      if (operatorType.indexOf('gte') > -1) {
        pre = '>= ';
      } else {
        pre = '>';
      }
    }
  }

  return pre;
};

/**
 * Returns min, max values for Operators.has_number_between case
 */
export const getMinMax = (args: string[] = []): { min: string; max: string } => {
  let min = '';
  let max = '';
  if (args.length) {
    if (strContainsNum(args[0])) {
      min = args[0];
    }
    if (args[1]) {
      if (strContainsNum(args[1])) {
        max = args[1];
      }
    }
  }
  return {
    min,
    max
  };
};

export const isFloat = (val: string): boolean => /^[+-]?\d?(\.\d*)?$/.test(val.trim());

export const isInt = (val: string): boolean => /^[+-]?\d+$/.test(val.trim());

export const strContainsNum = (str: string): boolean => {
  const trimmed = str.trim();
  if (isFloat(trimmed)) {
    return true;
  } else if (isInt(trimmed)) {
    return true;
  } else {
    return false;
  }
};

export const parseNum = (str: string): number => {
  const trimmed = (str || '').trim();
  if (isFloat(trimmed)) {
    return parseFloat(str);
  } else if (isInt(trimmed)) {
    return parseInt(trimmed, 10);
  }
};

export const validateCase = (keys: {
  operatorConfig: Operator;
  argument?: string;
  state?: string;
  district?: string;
  min?: string;
  max?: string;
  confidence?: string;
  intent?: SelectOption;
  exitName?: string;
  exitEdited?: boolean;
  classifier?: Asset;
}): Partial<CaseElementState> => {
  // when the exit is set, our arguments become required
  const validators = keys.exitEdited && keys.exitName ? [Required] : [];

  const updates: Partial<CaseElementState> = {
    operatorConfig: keys.operatorConfig
  };

  updates.district = { value: '', validationFailures: [] };
  updates.state = { value: '', validationFailures: [] };
  updates.min = { value: '', validationFailures: [] };
  updates.max = { value: '', validationFailures: [] };
  updates.argument = { value: '', validationFailures: [] };
  updates.intent = { value: null, validationFailures: [] };
  updates.confidence = { value: '', validationFailures: [] };

  if (keys.operatorConfig.operands > 0) {
    switch (keys.operatorConfig.type) {
      case Operators.has_number_eq:
      case Operators.has_number_gt:
      case Operators.has_number_gte:
      case Operators.has_number_lt:
      case Operators.has_number_lte:
        validators.push(NumOrExp);
        break;
      case Operators.has_date_eq:
      case Operators.has_date_lt:
      case Operators.has_date_gt:
        validators.push(Numeric);
        break;
    }

    if (keys.operatorConfig.type === Operators.has_number_between) {
      const max = keys.max || '';
      const min = keys.min || '';

      const maxExpression = max.indexOf('@') > -1;
      const minExpression = min.indexOf('@') > -1;
      const hasExpression = maxExpression || minExpression;

      const numeric = [Numeric];

      updates.min = validate(
        i18n.t('forms.minimum_value', 'Minimum value'),
        min,
        validators
          .concat(!minExpression ? numeric : [])
          .concat(
            !hasExpression
              ? [LessThan(parseFloat(keys.max), i18n.t('forms.the_maximum', 'the maximum'))]
              : []
          )
      );

      updates.max = validate(
        i18n.t('forms.maximum_value', 'Maximum value'),
        max,
        validators
          .concat(!maxExpression ? numeric : [])
          .concat(
            !hasExpression
              ? [MoreThan(parseFloat(keys.min), i18n.t('forms.the_minimum', 'the minimum'))]
              : []
          )
      );
    } else if (keys.operatorConfig.type === Operators.has_district) {
      updates.argument = validate(
        i18n.t('forms.state', 'State'),
        keys.argument || '',
        validators.concat([])
      );
    } else if (keys.operatorConfig.type === Operators.has_ward) {
      updates.state = validate(
        i18n.t('forms.state', 'State'),
        keys.state || '',
        validators.concat([])
      );
      updates.district = validate(
        i18n.t('forms.district', 'District'),
        keys.district || '',
        validators.concat([])
      );
    } else if (
      keys.operatorConfig.type === Operators.has_top_intent ||
      keys.operatorConfig.type === Operators.has_intent
    ) {
      const intentValidators = [IsValidIntent(keys.classifier)];
      if (keys.confidence) {
        intentValidators.push(Required);
      }
      updates.intent = validate(i18n.t('forms.intent', 'Intent'), keys.intent, intentValidators);
      updates.confidence = validate(
        i18n.t('forms.confidence', 'Confidence'),
        keys.confidence || '',
        validators.concat(keys.intent ? [Numeric, Required] : [Numeric])
      );
    } else {
      updates.argument = validate('Value', keys.argument || '', validators);
    }
  }

  updates.categoryNameEdited = !!keys.exitEdited;
  updates.categoryName = validate(
    i18n.t('forms.category', 'Category'),
    updates.categoryNameEdited ? keys.exitName : getCategoryName(updates),
    updates.argument.value ||
      (updates.min.value && updates.max.value) ||
      (updates.state.value && updates.district.value)
      ? [Required]
      : []
  );

  updates.valid =
    updates.state.validationFailures.length === 0 &&
    updates.district.validationFailures.length === 0 &&
    updates.min.validationFailures.length === 0 &&
    updates.max.validationFailures.length === 0 &&
    updates.argument.validationFailures.length === 0 &&
    updates.intent.validationFailures.length === 0 &&
    updates.confidence.validationFailures.length === 0 &&
    updates.categoryName.validationFailures.length === 0;

  return updates;
};

export const getCategoryName = (state: Partial<CaseElementState>): string => {
  if (state.categoryNameEdited) {
    return state.categoryName.value;
  }

  if (state.operatorConfig.operands === 0) {
    return state.operatorConfig.categoryName;
  }

  if (
    state.operatorConfig.type === Operators.has_intent ||
    state.operatorConfig.type === Operators.has_top_intent
  ) {
    if (state.intent.value) {
      return titleCase(state.intent.value.name.replace('_', ' '));
    }
  }

  if (
    state.operatorConfig.type === Operators.has_number_between &&
    state.min.value &&
    state.max.value
  ) {
    return `${state.min.value} - ${state.max.value}`;
  }

  if (isRelativeDate(state.operatorConfig.type)) {
    const count = parseInt(state.argument.value, 10);
    if (!isNaN(count)) {
      const today =
        state.operatorConfig.type === Operators.has_date_eq
          ? i18n.t('forms.today_proper', 'Today')
          : i18n.t('forms.today', 'today');
      const op = count < 0 ? ' - ' : ' + ';
      const inDays =
        ' ' + (Math.abs(count) === 1 ? i18n.t('forms.day', 'day') : i18n.t('forms.days', 'days'));
      return prefix(state.operatorConfig.type) + today + op + Math.abs(count) + inDays;
    }
  }

  if (state.argument && state.argument.value) {
    const pre = prefix(state.operatorConfig.type);
    const words = state.argument.value.match(/\w+/g);

    if (words && words.length > 0) {
      const [firstWord] = words;
      return pre + titleCase(firstWord);
    }

    return pre + titleCase(state.argument.value);
  }

  return '';
};
