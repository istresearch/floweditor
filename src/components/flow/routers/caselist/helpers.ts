import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { Operator } from 'config/interfaces';
import { createUUID } from 'utils';

export const createEmptyCase = (operator: Operator): CaseProps => {
  const uuid = createUUID();
  return {
    uuid,
    kase: {
      uuid,
      type: operator.type,
      arguments: [''],
      category_uuid: null
    },
    categoryName: '',
    valid: true
  };
};
