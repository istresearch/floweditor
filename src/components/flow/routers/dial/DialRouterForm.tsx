import { default as bindCallbacks } from 'auto-bind/react';
import * as React from 'react';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from 'components/flow/routers/dial/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import { FormState, StringEntry } from 'store/nodeEditor';
import { Alphanumeric, Required, StartIsNonNumeric, validate } from 'store/validators';
import i18n from 'config/i18n';

export interface DialRouterFormState extends FormState {
  phone: StringEntry;
  resultName: StringEntry;
}

export default class DialRouterForm extends React.Component<RouterFormProps, DialRouterFormState> {
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(this.props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  private handlePhoneUpdated(value: string): void {
    const phone = validate(i18n.t('forms.phone_number', 'Phone Number'), value, [Required]);
    this.setState({
      phone: phone,
      valid: this.state.valid && !hasErrors(phone)
    });
  }

  private handleUpdateResultName(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    if (this.state.valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  public renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p>Enter the phone number to dial</p>
        <TextInputElement
          name="phone"
          placeholder={i18n.t('forms.phone_number', 'Phone Number')}
          showLabel={false}
          autocomplete={true}
          onChange={this.handlePhoneUpdated}
          entry={this.state.phone}
        />
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
