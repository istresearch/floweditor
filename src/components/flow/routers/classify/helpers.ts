import {
  hasCases,
  createCaseProps,
  resolveRoutes,
  ResolvedRoutes,
  createRenderNode
} from 'components/flow/routers/helpers';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Types, Operators, VISIBILITY_HIDDEN } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { CallClassifier, SwitchRouter, Case, Exit, Category, RouterTypes } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry, FormEntry } from 'store/nodeEditor';
import { createUUID, scalarArrayEquals, snakify } from 'utils';
import { ClassifyRouterFormState } from 'components/flow/routers/classify/ClassifyRouterForm';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { getOperatorConfig } from 'config';

export const getOriginalAction = (settings: NodeEditorSettings): CallClassifier => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 && settings.originalNode.node.actions[0]);

  if (action.type === Types.call_classifier) {
    return action as CallClassifier;
  }
};

export const nodeToState = (settings: NodeEditorSettings): ClassifyRouterFormState => {
  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: 'Result' };
  let initialCases: CaseProps[] = [];

  let operand = DEFAULT_OPERAND;
  let classifier: FormEntry = { value: null };

  let hiddenCases: CaseProps[] = [];

  if (getType(settings.originalNode) === Types.split_by_intent) {
    const router = settings.originalNode.node.router as SwitchRouter;

    if (hasCases(settings.originalNode.node)) {
      initialCases = createCaseProps(router.cases, settings.originalNode);

      hiddenCases = initialCases.filter(
        (kase: CaseProps) => getOperatorConfig(kase.kase.type).visibility === VISIBILITY_HIDDEN
      );

      initialCases = initialCases.filter(
        (kase: CaseProps) => getOperatorConfig(kase.kase.type).visibility !== VISIBILITY_HIDDEN
      );
    }

    resultName = { value: router.result_name || 'Result' };
    const action = getOriginalAction(settings) as CallClassifier;
    classifier = { value: action.classifier };
    operand = action.input;
  }

  const state: ClassifyRouterFormState = {
    hiddenCases,
    resultName,
    classifier,
    operand: { value: operand },
    cases: initialCases,
    valid: true
  };

  return state;
};

export interface Route {
  type: Operators;
  arguments: string[];
  name: 'Other';
}

export const ensureRoute = (routes: ResolvedRoutes, route: Route) => {
  const existingCasePosition = routes.cases.findIndex(
    kase => kase.type === route.type && scalarArrayEquals(kase.arguments, route.arguments)
  );

  // if it already exists, make sure it's at the end
  if (existingCasePosition >= 0) {
    // already in the right place we are done
    if (existingCasePosition === routes.cases.length - 1) {
      return;
    }

    const [existingCase] = routes.cases.splice(existingCasePosition, 1);
    routes.cases.push(existingCase);

    return;
  }

  const exit: Exit = {
    uuid: createUUID()
  };

  const category: Category = {
    uuid: createUUID(),
    name: route.name,
    exit_uuid: exit.uuid
  };

  // otherwise let's add it
  const kase: Case = {
    uuid: createUUID(),
    type: route.type,
    arguments: route.arguments,
    category_uuid: category.uuid
  };

  routes.categories.push(category);
  routes.cases.push(kase);

  // our new exit should go right in front of the default exit
  routes.exits.splice(routes.exits.length - 1, 0, exit);
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ClassifyRouterFormState
): RenderNode => {
  let uuid = createUUID();

  const routes = resolveRoutes(
    [...state.cases, ...state.hiddenCases],
    false,
    settings.originalNode.node,
    'Failure'
  );

  // make sure we have an other route since failure is our default
  ensureRoute(routes, {
    type: Operators.has_category,
    arguments: ['Success', 'Skipped'],
    name: 'Other'
  });

  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const routerResultName = state.resultName.value;
  const actionResultName = '_' + routerResultName + ' Classification';

  const newAction: CallClassifier = {
    uuid,
    type: Types.call_classifier,
    result_name: actionResultName,
    input: state.operand.value,
    classifier: {
      uuid: state.classifier.value.uuid,
      name: state.classifier.value.name
    }
  };

  const router: SwitchRouter = {
    cases: routes.cases,
    operand: `@results.${snakify(actionResultName)}`,
    categories: routes.categories,
    type: RouterTypes.switch,
    default_category_uuid: routes.defaultCategory,
    result_name: routerResultName
  };

  return createRenderNode(
    settings.originalNode.node.uuid,
    router,
    routes.exits,
    Types.split_by_intent,
    [newAction]
  );
};
