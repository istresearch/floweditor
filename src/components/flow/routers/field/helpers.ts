import { NAME_PROPERTY } from 'components/flow/props';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { FieldRouterFormState } from 'components/flow/routers/field/FieldRouterForm';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes
} from 'components/flow/routers/helpers';
import { getContactProperties } from 'components/helpers';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { FlowTypes, Types } from 'config/interfaces';
import { getType, Scheme, SCHEMES } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';

export const getRoutableFields = (flowType: FlowTypes = null): Asset[] => {
  return [
    ...getContactProperties(flowType),
    ...SCHEMES.map((scheme: Scheme) => ({
      name: scheme.path,
      id: scheme.scheme,
      type: AssetType.Scheme
    }))
  ];
};

export const nodeToState = (
  settings: NodeEditorSettings,
  assetStore: AssetStore
): FieldRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: '' };

  let field: any = null;

  const type = getType(settings.originalNode);
  if (settings.originalNode && type === Types.split_by_contact_field) {
    const router = settings.originalNode.node.router as SwitchRouter;

    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }

    const operand = settings.originalNode.ui.config.operand;

    if (assetStore.fields) {
      if (operand.id in assetStore.fields.items) {
        const name = assetStore.fields.items[operand.id].name;
        field = { key: operand.id, label: name, type: operand.type };
      }
    }

    // couldn't find the asset, check our routable fields
    if (!field) {
      field = getRoutableFields().find((asset: Asset) => asset.id === operand.id);
    }
  }

  // our default is name
  if (!field) {
    field = NAME_PROPERTY;
  }

  return {
    cases: initialCases,
    resultName,
    field: { value: field },
    valid: true
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: FieldRouterFormState
): RenderNode => {
  const { cases, exits, defaultCategory: defaultExit, caseConfig, categories } = resolveRoutes(
    state.cases,
    false,
    settings.originalNode.node
  );

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  let operand = DEFAULT_OPERAND;

  const asset = state.field.value;

  let operandConfig = {
    id: asset.id,
    type: asset.type,
    name: asset.name
  };

  if (asset.type === AssetType.Scheme) {
    operand = `@(default(urn_parts(urns.${asset.id}).path, ""))`;
  } else if (asset.type === AssetType.ContactProperty) {
    operand = `@contact.${asset.id}`;
  } else if (asset.key) {
    operand = `@fields.${asset.key}`;
    operandConfig = {
      id: asset.key,
      name: asset.label,
      type: AssetType.Field
    };
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultExit,
    cases,
    categories,
    operand,
    ...optionalRouter
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.split_by_contact_field,
    [],
    {
      operand: operandConfig,
      cases: caseConfig
    }
  );

  return newRenderNode;
};
