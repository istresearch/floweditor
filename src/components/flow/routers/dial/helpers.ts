import { DialRouterFormState } from 'components/flow/routers/dial/DialRouterForm';
import { createRenderNode, getSwitchRouter } from 'components/flow/routers/helpers';
import { DIAL_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import {
  Case,
  Category,
  DialCategoryNames,
  DialStatus,
  Exit,
  Router,
  RouterTypes,
  SwitchRouter,
  Wait,
  WaitTypes
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';

export const nodeToState = (settings: NodeEditorSettings): DialRouterFormState => {
  let phone = '';
  let resultName: StringEntry = { value: '' };

  if (settings.originalNode && getType(settings.originalNode) === Types.wait_for_dial) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      phone = router.wait.phone;
      resultName = { value: router.result_name || '' };
    }
  }

  return {
    phone: { value: phone },
    resultName,
    valid: true
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: DialRouterFormState
): RenderNode => {
  let categories: Category[] = [];
  let cases: Case[] = [];
  let exits: Exit[] = [];

  const previousRouter = settings.originalNode ? getSwitchRouter(settings.originalNode.node) : null;

  // see if we are editing an existing dial router so we reuse exits
  if (previousRouter && previousRouter.wait && previousRouter.wait.type === WaitTypes.dial) {
    previousRouter.cases.forEach(kase => cases.push(kase));
    previousRouter.categories.forEach(category => categories.push(category));
    settings.originalNode.node.exits.forEach((exit: any) => exits.push(exit));
  } else {
    // otherwise, let's create some new ones
    exits = [
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      }
    ];

    categories = [
      {
        uuid: createUUID(),
        name: DialCategoryNames.Answered,
        exit_uuid: exits[0].uuid
      },
      {
        uuid: createUUID(),
        name: DialCategoryNames.NoAnswer,
        exit_uuid: exits[1].uuid
      },
      {
        uuid: createUUID(),
        name: DialCategoryNames.Busy,
        exit_uuid: exits[2].uuid
      },
      {
        uuid: createUUID(),
        name: DialCategoryNames.Failure,
        exit_uuid: exits[3].uuid
      }
    ];

    cases = [
      {
        uuid: createUUID(),
        type: Operators.has_only_text,
        arguments: [DialStatus.answered],
        category_uuid: categories[0].uuid
      },
      {
        uuid: createUUID(),
        type: Operators.has_only_text,
        arguments: [DialStatus.noAnswer],
        category_uuid: categories[1].uuid
      },
      {
        uuid: createUUID(),
        type: Operators.has_only_text,
        arguments: [DialStatus.busy],
        category_uuid: categories[2].uuid
      }
    ];
  }

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const wait = { type: WaitTypes.dial, phone: state.phone.value } as Wait;

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    wait: wait,
    default_category_uuid: categories[categories.length - 1].uuid,
    categories,
    cases,
    operand: DIAL_OPERAND,
    ...optionalRouter
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.wait_for_dial,
    []
  );

  return newRenderNode;
};
