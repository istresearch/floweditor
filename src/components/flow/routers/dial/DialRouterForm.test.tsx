import DialRouterForm from 'components/flow/routers/dial/DialRouterForm';
import { Types } from 'config/interfaces';
import { WaitTypes } from 'flowTypes';
import * as React from 'react';
import { fireEvent, render, getUpdatedNode, fireChangeText } from 'test/utils';
import { mock } from 'testUtils';
import { createDialRouter, getRouterFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { getSwitchRouter } from '../helpers';

const routerNode = createDialRouter('0979123456', 'Dial Result');
routerNode.ui = {
  position: { left: 0, top: 0 },
  type: Types.wait_for_dial
};

const routerProps = getRouterFormProps(routerNode);

mock(utils, 'createUUID', utils.seededUUIDs());

describe(DialRouterForm.name, () => {
  it('should render', () => {
    const { baseElement } = render(<DialRouterForm {...routerProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should reuse ids on updates', () => {
    const { getByText } = render(<DialRouterForm {...routerProps} />);
    fireEvent.click(getByText('Ok'));

    expect(getUpdatedNode(routerProps).node).toEqual(routerProps.nodeSettings.originalNode.node);
  });

  it('should update wait when phone number is changed', () => {
    const { getByTestId, getByText } = render(<DialRouterForm {...routerProps} />);

    fireChangeText(getByTestId('phone'), '@fields.supervisor_phone');
    fireEvent.click(getByText('Ok'));

    const router = getSwitchRouter(getUpdatedNode(routerProps).node);
    expect(router.wait.type).toBe(WaitTypes.dial);
    expect(router.wait.phone).toBe('@fields.supervisor_phone');
    expect(routerProps.updateRouter).toMatchCallSnapshot();
  });
});
