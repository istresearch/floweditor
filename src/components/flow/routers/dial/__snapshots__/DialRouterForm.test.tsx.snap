// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialRouterForm should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header wait_for_dial"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Wait for Forwarded Call
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"wait_for_dial\\",\\"name\\":\\"Wait for Forwarded Call\\",\\"description\\":\\"Wait for forwarded call to someone else\\",\\"visibility\\":[\\"voice\\"]}]"
              />
            </div>
          </div>
        </div>
        <p>
          Enter the phone number to dial
        </p>
        <div
          class="ele"
        >
          <div
            class="wrapper normal"
          >
            <input
              data-testid="phone"
              name="phone"
              placeholder="Phone Number"
              value="0979123456"
            />
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Dial Result"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.dial_result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DialRouterForm should update wait when phone number is changed 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
        },
        Object {
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
        Object {
          "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
        },
        Object {
          "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        },
        Object {
          "uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
        },
        Object {
          "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "answered",
            ],
            "category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
            "type": "has_only_text",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          Object {
            "arguments": Array [
              "no answer",
            ],
            "category_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
            "type": "has_only_text",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "arguments": Array [
              "busy",
            ],
            "category_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
          Object {
            "arguments": Array [
              "failed",
            ],
            "category_uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
            "type": "has_only_text",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "name": "Answered",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          Object {
            "exit_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "name": "No Answer",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
          Object {
            "exit_uuid": "6c22884b-0e35-4206-982e-18320691eda9",
            "name": "Busy",
            "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
          },
          Object {
            "exit_uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
            "name": "Failed",
            "uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
          },
          Object {
            "exit_uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
            "name": "Other",
            "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
          },
          Object {
            "exit_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "name": "No Response",
            "uuid": "f05fa107-03db-4315-a211-53da63603e6f",
          },
        ],
        "default_category_uuid": "f05fa107-03db-4315-a211-53da63603e6f",
        "operand": "@(default(resume.dial.status, \\"\\"))",
        "result_name": "Dial Result",
        "type": "switch",
        "wait": Object {
          "phone": "0979123456",
          "type": "dial",
        },
      },
      "uuid": "141e0356-9fdd-47d3-a12d-16efc5b45276",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "wait_for_dial",
    },
  },
]
`;
